cpp/src/transform/one_hot_encode.cu

```diff
@@ -0,0 +1,145 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/column/column_factories.hpp>
+#include <cudf/detail/copy.hpp>
+#include <cudf/detail/iterator.cuh>
+#include <cudf/table/row_operators.cuh>
+#include <cudf/types.hpp>
+#include <cudf/utilities/error.hpp>
+#include <cudf/utilities/traits.hpp>
+
+#include <thrust/iterator/counting_iterator.h>
+#include <thrust/transform.h>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <algorithm>
+
+namespace cudf {
+namespace detail {
+
+namespace {
+
+template <typename InputType, bool has_nulls>
+struct one_hot_encode_functor {
+  one_hot_encode_functor(column_device_view input, column_device_view category)
+    : _equality_comparator{input, category}, _input_size{input.size()}
+  {
+  }
+
+  bool __device__ operator()(size_type i)
+  {
+    size_type element_index  = i % _input_size;
+    size_type category_index = i / _input_size;
+    return _equality_comparator.template operator()<InputType>(element_index, category_index);
```

Comment:
These vars could be const, but also you could just inline the calculation. If you think it's more readable to name the vars first then just make them const.